src/lib/optimizer/optimizer.cpp

```diff
@@ -9,8 +9,8 @@
 
 namespace opossum {
 
-const Optimizer& Optimizer::get() {
-  static Optimizer optimizer{create_default_optimizer()};
+const std::shared_ptr<Optimizer>& Optimizer::get_default_optimizer() {
```

Comment:
More or less random place for comment:

> this will come in handy as one can now call sql_pipeline.set_optimizer(...). Need this for my MA.

If we now have different Optimizers, what is the point of having Singleton? There is no shared state, and each optimizer works independently of the others. If we now want a `DoesNothingOptimizer`, `OnlyJoinOptimizer` (or whatever) then I think we can get rid of the singleton. Or is there any reason this needs to stay a singleton? The `SQLTranslator` and `LQPTranslator` aren't Singeltons either.

I think a nicer solution (depending on how many Optimizers there should be in total) would be to keep the sttic methods for creating which hand out different Optimizer objects depending on the config.

```cpp
const Optimizer Optimizer::get_default_optimizer() { return Optimizer{rules}; }
const Optimizer Optimizer::get_dummy_optimizer() { return Optimizer{no-rules} }
const Optimizer Optimizer::get_super_fancy_optimizer( return Optimizer{super-fancy-rules} );
```
or just two constructors

```cpp
Optimizer::Optimizer() { _rules.emplace_back(default-rules); }
Optimizer::Optimizer(Rules rules) { _rules.emplace_back(rules); }
```

Also a shared_ptr singleton kinda conveys the wrong message, as nobody really owns it in the sense of a shared_ptr. With custom Optimizers however, this would be perfectly okay again :)

Reply:
The singleton was there to avoid creating the default Optimizer object over and over again - but you're right, for the moment there is no need.